### PR TITLE
PostgreSQL MVT path

### PR DESCRIPTION
**Pull request by dv343**
Adds a new path to generate MVT tiles without using Mapnik.

The new path needs PostGIS 2.4 since PostGIS is the one who generates the MVT tiles with the new path, with the St_AsMVT and St_AsMVTGeom functions.

Since complying with this requisite could be problematic, and we may want to rollback to the old functionality, a flag is used to switch between the old mapnik path and this new one.

The flag is `opts.mvt.usePgSQL`, and it is passed to the constructor by the upper layers. If it doesn't exists it will default to the old path.

Do not merge yet, since the tests are passing because they don't target the new functionality.

### CODE REVIEW

File: test/acceptance/torque.js [hunk 1]
```diff
@@ -293,10 +293,10 @@ describe('torque', function() {
         var testClient = new TestClient(mapConfig);
         testClient.getTile(0, 0, 0, {layer: 0, format: 'torque.json'}, function(err, torqueTile) {
             SubstitutionTokens.replace = replaceFn;
-
             assert.ok(!err, err);
-            assert.deepEqual(torqueTile, [{ x__uint8: 128, y__uint8: 128, vals__uint8: [2,3], dates__uint16: [1,0] }]);
-
+            torqueTile[0].vals__uint8.sort();
+            torqueTile[0].dates__uint16.sort();
+            assert.deepEqual(torqueTile, [{ x__uint8: 128, y__uint8: 128, vals__uint8: [2,3], dates__uint16: [0,1] }]);
```
Comment (rochoa) [inline test/acceptance/torque.js]: Why are these changes here?

File: lib/windshaft/backends/map_validator.js [hunk 2]
```diff
@@ -100,9 +100,9 @@ MapValidatorBackend.prototype.validate = function(mapConfigProvider, callback) {
                     var layerType = mapConfig.layerType(layerId);
 
                     if (layerType === 'mapnik') {
-                        if ( lyropt.interactivity ) {
-                            validateRemainingFormatLayerFns.push(validateMapnikGridJson(layerId));
-                        }
+                         if ( lyropt.interactivity ) {
```
Comment (rochoa) [inline lib/windshaft/backends/map_validator.js]: Can we avoid code style changes to keep the PR as clear as possible?

File: lib/windshaft/renderers/mapnik/geojson_renderer.js [hunk 3]
```diff
@@ -83,7 +83,11 @@ GeojsonRenderer.prototype._getTileData = function(cartoPSQL, coord, zoom, layer,
         // See https://github.com/mapnik/mapnik/wiki/ScaleAndPpi#scale-denominator
         scale_denominator: xyzResolution / 0.00028,
         pixel_width: xyzResolution,
-        pixel_height: xyzResolution
+        pixel_height: xyzResolution,
```
Comment (rochoa) [inline lib/windshaft/renderers/mapnik/geojson_renderer.js]: Why is this here?

File: lib/windshaft/renderers/pg_mvt/factory.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+///            http://github.com/CartoDB/CartoDB-SQL-API/blob/1.8.2/doc/API.md#json
+///            The 'readonly' parameter (false by default) requests
+///            that running the query should not allowed to change the database.
+///     - dbPoolParams: database connection pool params
+///          - size: maximum number of resources to create at any given time
+///          - idleTimeout: max milliseconds a resource can go unused before it should be destroyed
+///          - reapInterval: frequency to check for idle resources
+///
+function PgMvtFactory(options) {
+    // jshint newcap:false
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/factory.js]: Do we need this jshint?

File: lib/windshaft/renderers/pg_mvt/factory.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
+var _ = require('underscore');
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/factory.js]: Nowadays, it's more common to use `filename-with-dash-as-separator.js` format, could we change it here? 🙏

File: lib/windshaft/renderers/pg_mvt/factory.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+///          - size: maximum number of resources to create at any given time
+///          - idleTimeout: max milliseconds a resource can go unused before it should be destroyed
+///          - reapInterval: frequency to check for idle resources
+///
+function PgMvtFactory(options) {
+    // jshint newcap:false
+    this.options = options || {};
+    _.defaults(this.options, {
+        // TODO we most likely do not need such adaptor
+        sqlClass: PSQLAdaptor(PSQL, options.dbPoolParams)
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/factory.js]: Do we really need this `PSQLAdaptor` indirection? Couldn't we just use `PSQL` directly?

File: lib/windshaft/renderers/pg_mvt/factory.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+    getAdaptor: function (renderer, format, onTileErrorStrategy) {
+        return new BaseAdaptor(renderer, format, onTileErrorStrategy);
+    },
+
+    getRenderer: function (mapConfig, format, options, callback) {
+        var dbParams = RendererParams.dbParamsFromReqParams(options.params);
+        var layer = options.layer;
+
+        if (!formatToRenderer[format]) {
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/factory.js]: Wouldn't be better to call `this.supportsFormat`?

File: lib/windshaft/renderers/pg_mvt/factory.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+        var mapLayers = mapConfig.getLayers();
+        mapLayers.forEach((layer, layerIndex) => {
+            layer.id = mapConfig.getLayerId(layerIndex);
+        });
+
+
+        var mapnik_layers = [];
+        var layerFilter = options.layer;
+        if (typeof layerFilter === 'number') {
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/factory.js]: Why don't we rely on/implement this on `utils/layer_filter`?

File: lib/windshaft/renderers/pg_mvt/renderer.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                // See https://github.com/mapnik/mapnik/wiki/ScaleAndPpi#scale-denominator
+                scale_denominator: `(cdb_XYZ_Resolution(${z}) / 0.00028)`,
+                pixel_width: `cdb_XYZ_Resolution(${z})`,
+                pixel_height: `cdb_XYZ_Resolution(${z})`,
+                var_zoom: z,
+                var_x: x,
+                var_y: y
+            });
+
+            query += `coalesce((` +
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/renderer.js]: As we are already using template strings, I would avoid `+` concatenation when possible.

File: lib/windshaft/renderers/pg_mvt/renderer.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    /// @param z tile zoom
+    /// callback: will be called when done using nodejs protocol (err, data)
+    getTile: function (z, x, y, callback) {
+        var query = 'SELECT (';
+        this.layers.forEach((layer, i) => {
+            if (i >= 1) {
+                query += ' || ';
+            }
+            var nonGeoColumns = geojsonUtils.getGeojsonProperties(layer.options).join(',');
+            if (nonGeoColumns !== '') {
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/renderer.js]: I would put all columns, even the geometry one, in the array, so later you can just simply do `.join(',')`.

File: lib/windshaft/renderers/pg_mvt/renderer.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        this.sql(query, function (err, data) {
+            timer.end('query');
+            if (err) {
+                debug("Error running pg_mvt query " + query + ": " + err);
+                if (err.message) {
+                    err.message = "PgMvtRenderer: " + err.message;
+                }
+                callback(err);
+            } else {
+                callback(null, data.rows[0].st_asmvt, { 'Content-Type': 'application/x-protobuf' }, timer.getTimes());
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/renderer.js]: Should we sanity check `rows[0]. st_asmvt `?

File: lib/windshaft/renderers/renderer_factory.js [hunk 12]
```diff
@@ -29,7 +32,10 @@ function RendererFactory(opts) {
             opts.http.fallbackImage
         )
     ];
-    this.factories = availableFactories.reduce(function(factories, factory) {
+    if (opts.mvt.usePgSQL) {
```
Comment (rochoa) [inline lib/windshaft/renderers/renderer_factory.js]: Why not having it available all the time?

File: lib/windshaft/renderers/renderer_factory.js [hunk 13]
```diff
@@ -29,7 +32,10 @@ function RendererFactory(opts) {
...
+        availableFactories.push(new PgMvtRenderer.factory(opts.torque)); // TODO abusing torque
```
Comment (rochoa) [inline lib/windshaft/renderers/renderer_factory.js]: We cannot rely on torque options outside of torque.

File: test/acceptance/mvt.js [hunk 14]
```diff
@@ -4,13 +4,13 @@ var assert = require('../support/assert');
 var TestClient = require('../support/test_client');
 var mapnik = require('mapnik');
 
-describe('mvt', function() {
```
Comment (rochoa) [inline test/acceptance/mvt.js]: We are still testing with Mapnik implementation, aren't we?

File: test/acceptance/mvt.js [hunk 15]
```diff
@@ -132,35 +137,35 @@ describe('mvt', function() {
...
+            testClient.getTile(13, 4011, 3088, { layer: 'mapnik', format: 'mvt' }, multipleLayersValidation(done));
         });
 
-        it('invalid mvt layer', function(done) {
+        it('invalid mvt layer', function (done) {
             var testClient = new TestClient(mixedLayersMapConfig);
-            testClient.getTile(13, 4011, 3088, { layer: 0, format: 'mvt'}, function(err) {
+            testClient.getTile(13, 4011, 3088, { layer: 0, format: 'mvt' }, function (err) {
                 assert.ok(err);
-                assert.equal(err.message, 'Unsupported format mvt');
```
Comment (rochoa) [inline test/acceptance/mvt.js]: Why did we change the error message?

File: test/acceptance/mvt.js [hunk 16]
```diff
@@ -4,13 +4,13 @@ var assert = require('../support/assert');
 var TestClient = require('../support/test_client');
```
Comment (rochoa) [inline test/acceptance/mvt.js]: This file has too many changes related to code style. Could we revert all changes related to code style and fix them in another PR? It's hard to follow what it's actually changing.

File: test/acceptance/mvt.js [hunk 17]
```diff
@@ -4,13 +4,13 @@ var assert = require('../support/assert');
...
 var mapnik = require('mapnik');
 
-describe('mvt', function() {
+describe('mvt', function () {
```
Comment (rochoa) [inline test/acceptance/mvt.js]: I guess we need to configure this suite to run with both valued, true and false, for the usePgSQL configuration option, right?

File: test/acceptance/torque.js (hunk 1 above)
Comment (dv343) [reply test/acceptance/torque.js]: I will push these changes in a separated PR.

File: lib/windshaft/renderers/mapnik/geojson_renderer.js (hunk 3 above)
Comment (dv343) [reply lib/windshaft/renderers/mapnik/geojson_renderer.js]: I will remove those changes, but the geojson is going to be removed anyway in another PR.

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 5 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: What do you mean?

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 4 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: No, fixed

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 6 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: Maybe, that code is yours and Rafa's I haven't touched it. You know more than me.

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 7 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: Changed

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 8 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: I relied on it partially, just like mapnik renderer does. It's true that that code asks for a refactor, but IMO it's outside this PR goals, and it's not particularly related to MVT. See https://github.com/CartoDB/Windshaft/blob/master/lib/windshaft/renderers/mapnik/factory.js#L157

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 10 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: No, look at the query (https://github.com/CartoDB/Windshaft/pull/561/files/ff5ae340771e9fb180667128ad564d6614a3ff6b#diff-44ec6241c20bd52513803aa408451c69R56) Mixing columns would be a problem.

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 9 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: If you do that all the newlines and spaces will be sent to the DB, IMO, it's better to avoid that.

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 11 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: I don't think so. The query is already a workaround (for safety) for known PostGIS problems (which have been already addressed at PostGIS). If rows has zero length then the query is bugged or PostGIS behavior have changed, in any case the automatic exception will be Ok, IMO.

File: lib/windshaft/renderers/renderer_factory.js (hunk 12 above)
Comment (dv343) [reply lib/windshaft/renderers/renderer_factory.js]: For isolation, if the new path is disabled, then no code of the new path should be executed. If a bug is present in that part (at initialization time) then we don't need to deploy a fast fix, a configuration change would be enough.

Also, I think it's better to let pg_mvt renderer assume he has a PostGIS 2.4, if he hasn't then the factory shouldn't create him.

File: lib/windshaft/renderers/renderer_factory.js (hunk 13 above)
Comment (dv343) [reply lib/windshaft/renderers/renderer_factory.js]: I think this is related to the SQLAdapator, I think those options control DB connections. Again, I don't know about how we manage connections since I haven't programmed that part. It is also tightly coupled with windshaft-cartodb, so changing this requires a PR to windshaft-cartodb. I agree that using torque opts is ugly and should be fixed, but I need more information.

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 5 above)
Comment (rochoa) [reply lib/windshaft/renderers/pg_mvt/factory.js]: `lib/windshaft/renderers/pg_mvt/factory.js` -> `lib/windshaft/renderers/pg-mvt/factory.js`.

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 8 above)
Comment (rochoa) [reply lib/windshaft/renderers/pg_mvt/factory.js]: OK, fair enough.

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 10 above)
Comment (rochoa) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: Why?


```diff
diff --git a/lib/windshaft/renderers/pg_mvt/renderer.js b/lib/windshaft/renderers/pg_mvt/renderer.js
index 7726be26..6cf5ccdd 100644
--- a/lib/windshaft/renderers/pg_mvt/renderer.js
+++ b/lib/windshaft/renderers/pg_mvt/renderer.js
@@ -35,10 +35,8 @@ Renderer.prototype = {
             if (i >= 1) {
                 query += ' || ';
             }
-            var nonGeoColumns = geojsonUtils.getGeojsonProperties(layer.options).join(',');
-            if (nonGeoColumns !== '') {
-                nonGeoColumns += ',';
-            }
+            var columns = geojsonUtils.getGeojsonProperties(layer.options);
+            columns.push(`ST_AsMVTGeom(the_geom_webmercator, CDB_XYZ_Extent(${x},${y},${z}), ${this.mvt_extent}, 0, true)`);

             var subQuery = SubstitutionTokens.replace(layer.options.sql, {
                 bbox: `CDB_XYZ_Extent(${x},${y},${z})`,
@@ -53,7 +51,7 @@ Renderer.prototype = {

             query += `coalesce((` +
                 `select st_asmvt(geom, '${layer.id}') FROM` +
-                `(SELECT ${nonGeoColumns} ST_AsMVTGeom(the_geom_webmercator, CDB_XYZ_Extent(${x},${y},${z}), ${this.mvt_extent}, 0, true)` +
+                `(SELECT ${columns.join(',')}` +
                 `FROM (${subQuery}) AS cdbq where the_geom_webmercator && CDB_XYZ_Extent(${x},${y},${z}) ) as geom where geom is not null` +
                 `), '')`;
         });
```

File: test/acceptance/mvt.js (hunk 16 above)
Comment (dv343) [reply test/acceptance/mvt.js]: Changed

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 9 above)
Comment (rochoa) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: Is that a problem? The `subQuery` can contain anything, and it is also sent to the db.

File: lib/windshaft/backends/map_validator.js (hunk 2 above)
Comment (dv343) [reply lib/windshaft/backends/map_validator.js]: Changed

File: test/acceptance/mvt.js (hunk 14 above)
Comment (dv343) [reply test/acceptance/mvt.js]: No, we are not, see last pushed version

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 11 above)
Comment (rochoa) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: What do you mean by _automatic exception_?

File: lib/windshaft/renderers/renderer_factory.js (hunk 12 above)
Comment (rochoa) [reply lib/windshaft/renderers/renderer_factory.js]: I don't understand this. Making the factory available doesn't mean it has to be used, that's why the `renderer_factory.js` has:

```js
if (format === 'mvt' && this.opts.mvt.usePgSQL) {
    return 'pg_mvt';
}
```

File: lib/windshaft/renderers/renderer_factory.js (hunk 13 above)
Comment (rochoa) [reply lib/windshaft/renderers/renderer_factory.js]: It's not about the connections, but about the params needed to create the connection. If we need to change Windshaft-cartodb, that's OK. But from a library standpoint of view, we cannot say: "_Hey, for MVT we need the Torque options_", that's not clear. So let's fix this.

File: test/acceptance/torque.js (hunk 1 above)
Comment (dv343) [reply test/acceptance/torque.js]: Changed

File: lib/windshaft/backends/map_validator.js (hunk 2 above)
Comment (dv343) [reply lib/windshaft/backends/map_validator.js]: Changed

File: lib/windshaft/renderers/mapnik/geojson_renderer.js (hunk 3 above)
Comment (dv343) [reply lib/windshaft/renderers/mapnik/geojson_renderer.js]: Changed

File: lib/windshaft/renderers/pg_mvt/factory.js [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+///     - dbPoolParams: database connection pool params
+///          - size: maximum number of resources to create at any given time
+///          - idleTimeout: max milliseconds a resource can go unused before it should be destroyed
+///          - reapInterval: frequency to check for idle resources
+///
+function PgMvtFactory(options) {
+    this.options = options || {};
+    _.defaults(this.options, {
+        // TODO we most likely do not need such adaptor
+        sqlClass: new PSQLAdaptor(PSQL, options.dbPoolParams)
```
Comment (rochoa) [inline lib/windshaft/renderers/pg_mvt/factory.js]: My previous comment is missing. I would remove `PSQLAdaptor` and use `PSQL` directly. The `PSQLAdaptor` was part of torque renderer at the very beginning, but it doesn't make sense here anymore.

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 4 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: We don't need it, changed.

File: test/acceptance/mvt.js [hunk 19]
```diff
@@ -133,33 +146,33 @@ describe('mvt', function() {
...
+            var testClient = new TestClient(mixedLayersMapConfig, options);
             testClient.getTile(13, 4011, 3088, { layer: 'mapnik', format: 'mvt'}, multipleLayersValidation(done));
         });
 
         it('invalid mvt layer', function(done) {
-            var testClient = new TestClient(mixedLayersMapConfig);
+            var testClient = new TestClient(mixedLayersMapConfig, options);
             testClient.getTile(13, 4011, 3088, { layer: 0, format: 'mvt'}, function(err) {
                 assert.ok(err);
-                assert.equal(err.message, 'Unsupported format mvt');
```
Comment (rochoa) [inline test/acceptance/mvt.js]: My previous comment is missing: Why did we change the error message?

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 10 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: You are mixing now raw column names with SQL expressions, but Ok.

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 5 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: Ok, I will change it.

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 9 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: Changed

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 11 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: I mean that JS will raise an exception at that `[0].st_asmvt` indirection, right?

File: lib/windshaft/renderers/renderer_factory.js (hunk 12 above)
Comment (dv343) [reply lib/windshaft/renderers/renderer_factory.js]: Invoking the constructor is using it, and having it available is invoking the constructor.

File: lib/windshaft/renderers/renderer_factory.js (hunk 13 above)
Comment (dv343) [reply lib/windshaft/renderers/renderer_factory.js]: It would be Ok to set all those params at `renderer.dbparams` (in terms of windshaft-cartodb)?

File: lib/windshaft/renderers/renderer_factory.js (hunk 12 above)
Comment (rochoa) [reply lib/windshaft/renderers/renderer_factory.js]: You are just making the factory available. It not a heavy class, and it's done just once.

File: lib/windshaft/renderers/pg_mvt/renderer.js (hunk 11 above)
Comment (rochoa) [reply lib/windshaft/renderers/pg_mvt/renderer.js]: It will, but you are not handling it. And you are in async context, meaning that the _client_ expecting to be called back won't get anything.

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 18 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: My comment is missing too. That code is yours and Rafa's, I don't know about it, what was its purpose? Why is it safe to remove it?

File: test/acceptance/mvt.js (hunk 19 above)
Comment (dv343) [reply test/acceptance/mvt.js]: I think that including the invalid format gives more information.

File: lib/windshaft/renderers/pg_mvt/factory.js (hunk 5 above)
Comment (dv343) [reply lib/windshaft/renderers/pg_mvt/factory.js]: Changed

File: lib/windshaft/renderers/renderer_factory.js (hunk 12 above)
Comment (dv343) [reply lib/windshaft/renderers/renderer_factory.js]: Changed

File: lib/windshaft/renderers/pg-mvt/factory.js [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+///     - dbPoolParams: database connection pool params
+///          - size: maximum number of resources to create at any given time
+///          - idleTimeout: max milliseconds a resource can go unused before it should be destroyed
+///          - reapInterval: frequency to check for idle resources
+///
+function PgMvtFactory(options) {
+    this.options = options || {};
+    _.defaults(this.options, {
+        // TODO we most likely do not need such adaptor
+        sqlClass: new PSQLAdaptor(PSQL, options.dbPoolParams)
```
Comment (rafatower) [inline lib/windshaft/renderers/pg-mvt/factory.js]: Instead of using the torque `PSQLAdaptor` I'd try to replace it by `PSQL` or see if we really need anything specific from it (I don't think so) and  replace it by its own adaptor, if needed.

File: lib/windshaft/renderers/renderer_factory.js [hunk 21]
```diff
@@ -13,6 +14,8 @@ function RendererFactory(opts) {
     opts.http = opts.http || {};
     opts.mapnik = opts.mapnik || {};
     opts.torque = opts.torque || {};
+    opts.mvt = opts.mvt || {};
+    this.opts = opts;
```
Comment (rafatower) [inline lib/windshaft/renderers/renderer_factory.js]: why is this line required?

File: lib/windshaft/renderers/renderer_factory.js [hunk 22]
```diff
@@ -27,9 +30,10 @@ function RendererFactory(opts) {
             opts.http.timeout,
             opts.http.proxy,
             opts.http.fallbackImage
-        )
+        ),
+        new PgMvtRenderer.factory(opts.torque)
```
Comment (rafatower) [inline lib/windshaft/renderers/renderer_factory.js]: I think this was already commented but yet: are we relying on torque options? I think it should stick to `opts.mvt`

File: lib/windshaft/renderers/renderer_factory.js [hunk 23]
```diff
@@ -86,7 +90,11 @@ function genericMakeRenderer(factory, mapConfig, params, limits, onTileErrorStra
     );
 }
 
-function getFactoryName(mapConfig, layer, format) {
+RendererFactory.prototype.getFactoryName = function (mapConfig, layer, format) {
+    if (this.opts.mvt.usePostGIS && format === 'mvt' && isMapnikFactory(mapConfig, layer, format)) {
```
Comment (rafatower) [inline lib/windshaft/renderers/renderer_factory.js]: why is the `isMapnikFactory` check needed?

File: lib/windshaft/renderers/pg-mvt/factory.js [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+    supportsFormat: function (format) {
+        return !!formatToRenderer[format];
+    },
+
+    getAdaptor: function (renderer, format, onTileErrorStrategy) {
+        return new BaseAdaptor(renderer, format, onTileErrorStrategy);
+    },
+
+    getRenderer: function (mapConfig, format, options, callback) {
+        // jshint maxcomplexity: 7
```
Comment (rochoa) [inline lib/windshaft/renderers/pg-mvt/factory.js]: BTW, we should not introduce this in new code.

File: lib/windshaft/renderers/renderer_factory.js (hunk 21 above)
Comment (dv343) [reply lib/windshaft/renderers/renderer_factory.js]: Because `getFactoryName` needs to know if the PostGIS path is enabled

File: lib/windshaft/renderers/renderer_factory.js (hunk 22 above)
Comment (dv343) [reply lib/windshaft/renderers/renderer_factory.js]: Yes, we should! I didn't see this reference, good catch

File: lib/windshaft/renderers/renderer_factory.js (hunk 23 above)
Comment (dv343) [reply lib/windshaft/renderers/renderer_factory.js]: Well... this comes from a long discussion with @rochoa The way I see it we had 2 options: being consistent or doing it "right". They are mutually exclusive, and we decided to go with consistent. That function takes care of filtering.

Comment (rochoa) [reply lib/windshaft/renderers/renderer_factory.js]: MVT format only make sense for layers where you have an SQL option (type:mapnik|cartodb). With this, we avoid trying to generate MVT tiles from `plain` or `http` layers, where it does not make sense at all.

File: lib/windshaft/renderers/pg-mvt/factory.js (hunk 20 above)
Comment (dv343) [reply lib/windshaft/renderers/pg-mvt/factory.js]: I've just removed it

File: lib/windshaft/renderers/pg-mvt/factory.js (hunk 24 above)
Comment (dv343) [reply lib/windshaft/renderers/pg-mvt/factory.js]: Fixed

File: .travis.yml [hunk 25]
```diff
@@ -1,21 +1,13 @@
...
-  - NPROCS=1 JOBS=1 PGUSER=postgres CXX=g++-4.9
+services:
+  - docker
 
-language: node_js
-node_js:
-  - "6"
-dist: precise
+before_install:
+  - docker pull dmanzanares/windshaft-testing
```
Comment (rafatower) [inline .travis.yml]: I don't feel 100% comfortable downloading from anyone's personal docker hubs (or any other repository). Don't know if there's any quick solution for this

File: Dockerfile [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+# Configure PostgreSQL
+RUN echo "local     all       all                     trust" > /etc/postgresql/9.5/main/pg_hba.conf
+RUN echo "host      all       all       0.0.0.0/0     trust" >> /etc/postgresql/9.5/main/pg_hba.conf
+RUN echo "listen_addresses='*'" >> /etc/postgresql/9.5/main/postgresql.conf
+
+RUN apt-get -y purge $DEV_DEPENDENCIES
+RUN apt-get -y autoremove
+
+WORKDIR /srv
+CMD /etc/init.d/postgresql start && node -v && export NPROCS=1 && export JOBS=1 && export CXX=g++-4.9 && npm install && export PGUSER=postgres && npm test
```
Comment (rafatower) [inline Dockerfile]: :clap: :clap: vamos!

File: lib/windshaft/renderers/renderer_factory.js (hunk 23 above)
Comment (rafatower) [reply lib/windshaft/renderers/renderer_factory.js]: understood, thanks for the clarification.

**Comment by dv343** [comment]
From my point of view this branch is Ok now. However, it needs PostGIS 2.4 to use the new MVT path. If PostGIS 2.4 is not present the `opts` param should reflect that by disabling the new path.

The tests are ran in a docker xenial environment that compiles PostGIS 2.4 from sources.

File: lib/windshaft/renderers/pg-mvt/renderer.js [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+            });
+
+            query += `(
+                select st_asmvt(geom, '${layer.id}') FROM
+                (SELECT ${nonGeoColumns}
+                    ST_AsMVTGeom(the_geom_webmercator, CDB_XYZ_Extent(${x},${y},${z}), ${this.mvt_extent}, 0, true)
+                FROM
+                    (${subQuery}) AS cdbq where the_geom_webmercator && CDB_XYZ_Extent(${x},${y},${z}) )
+                    AS geom
+                )`;
```
Comment (rafatower) [inline lib/windshaft/renderers/pg-mvt/renderer.js]: pretty neat after removing the workarounds :)

File: .travis.yml (hunk 25 above)
Comment (dv343) [reply .travis.yml]: That's true, and it doesn't seem very professional... We should have a Docker organization with permissions and all that stuff... But at least the Dockerfile used to generate the image is in the repo.

File: lib/windshaft/renderers/pg-mvt/renderer.js (hunk 27 above)
Comment (dv343) [reply lib/windshaft/renderers/pg-mvt/renderer.js]: That's because the optimizations are missing :joy:

File: .travis.yml (hunk 25 above)
Comment (javisantana) [reply .travis.yml]: I've created a company account

**Comment by javisantana** [comment]
what does this PR needs to be merged?

**Comment by dv343** [comment]
@javisantana 
It needs a refactor to fix a bug in the filtering: https://github.com/CartoDB/Windshaft/pull/564 and partially merge https://github.com/CartoDB/Windshaft/tree/mvt-bug to check that we are downloading only those columns.
